protocols/gossipsub/src/layer.rs

```diff
@@ -0,0 +1,1788 @@
+        let mut iwant_ids = HashMap::new();
+
+        for (topic, ids) in ihave_msgs {
+            // only process the message if we are subscribed
+            if !self.mesh.contains_key(&topic) {
+                info!(
+                    "IHAVE: Ignoring IHAVE - Not subscribed to topic: {:?}",
+                    topic
+                );
+                return; // continue
+            }
+
+            for id in ids {
+                if !self.received.contains(&id) {
+                    // have not seen this message, request it
+                    iwant_ids.insert(id, true);
+                }
+            }
+        }
+
+        if !iwant_ids.is_empty() {
+            // Send the list of IWANT control messages
+            info!("IHAVE: Sending IWANT message");
+            self.events.push_back(NetworkBehaviourAction::SendEvent {
+                peer_id: peer_id.clone(),
+                event: GossipsubRpc {
+                    subscriptions: Vec::new(),
+                    messages: Vec::new(),
+                    control_msgs: vec![GossipsubControlAction::IWant {
+                        message_ids: iwant_ids.keys().cloned().collect(),
+                    }],
+                },
+            });
+        }
+        debug!("Completed IHAVE handling for peer: {:?}", peer_id);
+    }
+
+    /// Handles an IWANT control message. Checks our cache of messages. If the message exists it is
+    /// forwarded to the requesting peer.
+    fn handle_iwant(&mut self, peer_id: &PeerId, iwant_msgs: Vec<String>) {
+        debug!("Handling IWANT for peer: {:?}", peer_id);
+        // build a hashmap of available messages
+        let mut cached_messages = HashMap::new();
+
+        for id in iwant_msgs {
+            // if we have it, add it do the cached_messages mapping
+            if let Some(msg) = self.mcache.get(&id) {
+                cached_messages.insert(id.clone(), msg.clone());
+            }
+        }
+
+        if !cached_messages.is_empty() {
+            info!("IWANT: Sending cached messages to peer: {:?}", peer_id);
+            // Send the messages to the peer
+            let message_list = cached_messages.values().cloned().collect();
+            self.events.push_back(NetworkBehaviourAction::SendEvent {
+                peer_id: peer_id.clone(),
+                event: GossipsubRpc {
+                    subscriptions: Vec::new(),
+                    messages: message_list,
+                    control_msgs: Vec::new(),
+                },
+            });
+        }
+        debug!("Completed IWANT handling for peer: {:?}", peer_id);
+    }
+
+    /// Handles GRAFT control messages. If subscribed to the topic, adds the peer to mesh, if not, responds
+    /// with PRUNE messages.
+    fn handle_graft(&mut self, peer_id: &PeerId, topics: Vec<TopicHash>) {
+        debug!("Handling GRAFT message for peer: {:?}", peer_id);
+
+        let mut to_prune_topics = HashMap::new();
```

Comment:
Another `HashSet` maybe?

Reply:
Awesome, didn't realise `HashSet` existed. There are few places I will update with this. Thanks.